商品情報編集機能

Bakiyama opened the pull request:
# What
商品情報編集機能

# Why
商品の編集を可能にするため。
売却済みの場合の動作などは実装していません。

Bakiyama committed:
binding.pry導入

Bakiyama committed:
ルーティングとコントローラーを追加

Bakiyama committed:
詳細ページの「商品の編集」ボタンにeditのパスを指定

Bakiyama committed:
編集ページのビューファイルを作成

Bakiyama committed:
updateアクションのインスタンス変数を修正

norihiro9779 commented:
こちらのプルリクエストですが、実際の画面や機能ができているか確認させていただきたいので、下記キャプチャ画像をご提出いただけますか？
```
・商品編集に成功する動画
・商品編集に失敗し、エラーが表示される動画
・何も編集せずに更新をしても画像無しの商品にならない動画
・ログイン状態の出品者だけが商品情報編集ページに遷移できる動画
・ログイン状態の出品者以外のユーザーは、URLを直接入力して出品していない商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画
・ログアウト状態のユーザーは、URLを直接入力して商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画
・商品名やカテゴリーの情報など、すでに登録されている商品情報は商品情報編集画面を開いた時点で表示される動画（画像に関しては、表示されない状態で良い）
```
「Gyazo　使い方」とgoogleで検索し、ツールによって生成されたURLをこのPRに貼っていただければ幸いです。
その後、再度レビュー依頼フォームにてご提出をお願いいたします。

Bakiyama commented:
・商品編集に成功する動画
https://gyazo.com/033ac92d9cfe64f4f648562c7db0b3cb

・商品編集に失敗し、エラーが表示される動画
https://gyazo.com/e73f6c672d064a2243530e0486a3a48b

・何も編集せずに更新をしても画像無しの商品にならない動画
https://gyazo.com/b088f60c9bfe4c62a3e685df942016e1

・ログイン状態の出品者だけが商品情報編集ページに遷移できる動画
https://gyazo.com/911dd2170160ad19d0e964218719907a

・ログイン状態の出品者以外のユーザーは、URLを直接入力して出品していない商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/f2dde0ca86b26116de6f1251eb470e23

・ログアウト状態のユーザーは、URLを直接入力して商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画
https://gyazo.com/592aa22a09d00de42264b7ca8cc24990

・商品名やカテゴリーの情報など、すでに登録されている商品情報は商品情報編集画面を開いた時点で表示される動画（画像に関しては、表示されない状態で良い）
https://gyazo.com/f2a42b7229a5beb3fdc4ed91cd6954a2

chisesakai commented on app/controllers/items_controller.rb:
```diff
@@ -1,5 +1,5 @@
 class ItemsController < ApplicationController
-  before_action :authenticate_user!, only: [:new, :create]
+  before_action :authenticate_user!, only: [:new, :create, :edit]
```
こちら、商品出品機能同様、`update`アクションも追加しましょう！

chisesakai commented on app/controllers/items_controller.rb:
```diff
@@ -22,6 +22,22 @@ def create
   end
 
+  def edit
+    @item = Item.find(params[:id])
```
```ruby
    @item = Item.find(params[:id])
```
9,26,33行目で定義している@itemの記述が重複しているようです。
こちらはprivate以下にメソッド化して、beforeアクションで呼び出す形に修正しておきましょう。

複数回使用する変数をメソッド化することで可読性が上がり、改修作業が発生した際も手を加えやすくなります。

chisesakai commented on app/controllers/items_controller.rb:
```diff
@@ -22,6 +22,22 @@ def create
+    @item = Item.find(params[:id])
+    unless current_user.id == @item.user_id
+      redirect_to action: :index
+    end
```
```ruby
    unless current_user.id == @item.user_id
      redirect_to action: :index
    end
```
こちらは、`authenticate_user!`と同様、セキュリティ強化のためupdateアクションにも反映させましょう。

記述が重複しますので、こちらもメソッド化してbefore_actionで呼び出すのが良いですね。

chisesakai reviewed (commented):
レビュー依頼ありがとうございます。
キャプチャ拝見させていただきました。挙動を確認することが出来ました、ありがとうございます！
#### 3件コメントいたしましたので、ご確認をお願いいたします。

Bakiyama committed:
コントローラーを修正

・authenticate_userにupdateを追加

・@item = Item.find(params[:id])を
show, edit, updateのbefore_actionに設定。

・edit・updateアクションをするユーザーとアクセスしたユーザーのIDを確かめる条件式をbefore_actionに設定

norihiro9779 reviewed (commented):
レビュー依頼ありがとうございます。
丁寧にコードが記述されておりますね！
LGTMとさせていただきます！実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう！
